troposphere/__init__.py

```diff
@@ -162,10 +163,50 @@ def validate(self):
         pass
 
     @classmethod
-    def from_dict(cls, title, dict):
-        obj = cls(title)
-        obj.properties.update(dict)
-        return obj
+    def _from_dict(cls, title=None, **kwargs):
+        props = {}
+        for prop_name, value in kwargs.items():
+            try:
+                prop_attrs = cls.props[prop_name]
+            except KeyError:
+                raise AttributeError("Object type %s does not have a "
+                                     "%s property." % (cls.__name__,
+                                                       prop_name))
+            prop_type = prop_attrs[0]
+            if prop_name in kwargs:
+                value = kwargs[prop_name]
+                is_aws_object = False
+                try:
+                    is_aws_object = issubclass(prop_type, BaseAWSObject)
+                # prop_type isn't a class
+                except TypeError:
+                    pass
+                if is_aws_object:
+                    if not isinstance(value, collections.Mapping):
+                        raise ValueError("Property definition for %s must be "
+                                         "a Mapping type" % prop_name)
+                    value = prop_type._from_dict(**value)
+
+                if isinstance(prop_type, list):
+                    if not isinstance(value, list):
+                        raise TypeError("Attribute %s must be a "
+                                        "list." % prop_name)
+                    new_value = []
+                    for v in value:
+                        if not isinstance(v, collections.Mapping):
+                            raise ValueError(
```

Comment:
how would this work if this was a list of role names? does this always have to be a list of mapping types?